File: Java Extras/src/Snippets/LCS.java
package Snippets;

import lombok.Data;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertTrue;

/*
    Longest Common Sequence Solution
 */
public class LCS {

    @Data
    static class StringPair {
        final String first;
        final String second;
    }

    static Map<StringPair, String> cache = new HashMap<>();

    static String lcs(String a, String b) {
        if (a.length() == 0 || b.length() == 0) return "";

        StringPair current = new StringPair(a, b);
        if (cache.containsKey(current)) return cache.get(current);

        String result;
        if (a.charAt(a.length() - 1) == b.charAt(b.length() - 1)) {
            result = lcs(a.substring(0, a.length() - 1), b.substring(0, b.length() - 1)) + a.charAt(a.length() - 1);
        } else {
            String s1 = lcs(a.substring(0, a.length() - 1), b);
            String s2 = lcs(a, b.substring(0, b.length() - 1));
            result = s1.length() > s2.length() ? s1 : s2;
        }

        cache.put(current, result);
        return result;
    }

    // same but traversing the string forward in stead of backwards
    static String lcs2(String a, String b) {
        if (a.length() == 0 || b.length() == 0) return "";

        StringPair current = new StringPair(a, b);
        if (cache.containsKey(current)) return cache.get(current);

        String result;
        if (a.charAt(0) == b.charAt(0)) {
            result = a.charAt(0) + lcs2(a.substring(1), b.substring(1));
        } else {
            String s1 = lcs2(a.substring(1), b);
            String s2 = lcs2(a, b.substring(1));
            result = s1.length() > s2.length() ? s1 : s2;
        }

        cache.put(current, result);
        return result;
    }

    @Test
    static void doTestsLCS1() {
        assertTrue(Stream.of(
                (""         .equals(lcs("", ""))),
                ("final"    .equals(lcs("nothardlythefinaltest", "zzzfinallyzzz"))),
                ("abc"      .equals(lcs("abcdef", "abc"))),
                ("acf"      .equals(lcs("abcdef", "acf"))),
                ("nottest"  .equals(lcs("anothertest", "notatest"))),
                ("12356"    .equals(lcs("132535365", "123456789")))
        ).allMatch(it -> it));

        cache.clear();
    }

    @Test
    static void doTestsLCS2() {
        assertTrue(Stream.of(
                (""         .equals(lcs2("", ""))),
                ("final"    .equals(lcs2("nothardlythefinaltest", "zzzfinallyzzz"))),
                ("abc"      .equals(lcs2("abcdef", "abc"))),
                ("acf"      .equals(lcs2("abcdef", "acf"))),
                ("nottest"  .equals(lcs2("anothertest", "notatest"))),
                ("12356"    .equals(lcs2("132535365", "123456789")))
        ).allMatch(it -> it));

        cache.clear();
    }

    public static void main(String[] args) {
        doTestsLCS1();
        doTestsLCS2();
    }
}

File: Kotlin Tutorial + Snippets/src/snippets/LCS.kt
package snippets

import kotlin.collections.HashMap
import kotlin.test.assertTrue

val cache: MutableMap<Pair<String, String>, String> = HashMap()

fun lcs(a: String, b: String): String {
    if (a.isEmpty() || b.isEmpty()) return ""

    val current = Pair(a, b)
    if (cache.containsKey(current)) return cache[current].orEmpty()

    val result = if (a.last() == b.last()) {
        lcs(a.dropLast(1), b.dropLast(1)) + a.last()
    } else {
        val s1 = lcs(a.dropLast(1), b)
        val s2 = lcs(a, b.dropLast(1))
        if (s1.length > s2.length) s1 else s2
    }

    cache[current] = result
    return result
}

// ************************************************************************************
// pretty much same solution but going forwards instead of backwards through the string
val m = mutableMapOf<Pair<String, String>, String>()

fun lcs2(a: String, b: String): String {
    if (a.isEmpty() || b.isEmpty()) return ""
    if (a.first() == b.first()) return a.first() + lcs2(a.drop(1), b.drop(1))

    val s1 = m.getOrPut(a to b.drop(1)) { lcs2(a, b.drop(1)) }
    val s2 = m.getOrPut(a.drop(1) to b) { lcs2(a.drop(1), b) }
    return if (s1.length > s2.length) s1 else s2
}
// ************************************************************************************

fun doTests(lcsFunction: (String, String) -> String) {
    arrayOf(
            (""           == lcsFunction("", "")),
            ("final"      == lcsFunction("nothardlythefinaltest", "zzzfinallyzzz")),
            ("abc"        == lcsFunction("abcdef", "abc")),
            ("acf"        == lcsFunction("abcdef", "acf")),
            ("nottest"    == lcsFunction("anothertest", "notatest")),
            ("12356"      == lcsFunction("132535365", "123456789")),
    ).forEach(::assertTrue)
    cache.clear()
}

fun main() {
    doTests(::lcs)
    doTests(::lcs2)
}